generators/csharp/sdk/src/oauth/OauthTokenProviderGenerator.ts

```diff
@@ -0,0 +1,267 @@
+    private scheme: OAuthScheme;
+    private tokenEndpointHttpService: HttpService;
+    private tokenEndpointReference: EndpointReference;
+    private tokenEndpoint: HttpEndpoint;
+
+    private static readonly BUFFER_IN_MINUTES_FIELD = csharp.field({
+        access: "private",
+        const_: true,
+        name: "BufferInMinutes",
+        type: csharp.Type.double(),
+        initializer: csharp.codeblock("2")
+    });
+
+    private static readonly ACCESS_TOKEN_FIELD = csharp.field({
+        access: "private",
+        name: "_accessToken",
+        type: csharp.Type.optional(csharp.Type.string())
+    });
+
+    private static readonly EXPIRES_AT_FIELD = csharp.field({
+        access: "private",
+        name: "_expiresAt",
+        type: csharp.Type.optional(csharp.Type.dateTime())
+    });
+
+    private static readonly CLIENT_ID_FIELD = csharp.field({
+        access: "private",
+        name: "_clientId",
+        type: csharp.Type.string()
+    });
+
+    private static readonly CLIENT_SECRET_FIELD = csharp.field({
+        access: "private",
+        name: "_clientSecret",
+        type: csharp.Type.string()
+    });
+
+    constructor({ context, scheme }: OauthTokenProviderGenerator.Args) {
+        super(context);
+        this.scheme = scheme;
+        this.classReference = csharp.classReference({
+            namespace: this.context.getCoreNamespace(),
+            name: OAUTH_TOKEN_PROVIDER_CLASS_NAME
+        });
+
+        this.tokenEndpointReference = this.scheme.configuration.tokenEndpoint.endpointReference;
+        this.tokenEndpointHttpService = this.context.getHttpServiceOrThrow(this.tokenEndpointReference.serviceId);
+        const httpEndpoint = this.tokenEndpointHttpService.endpoints.find(
+            (endpoint) => endpoint.id === this.tokenEndpointReference.endpointId
+        );
+
+        if (httpEndpoint == null) {
+            throw new Error(`Failed to find token endpoint ${this.tokenEndpointReference.endpointId}`);
+        }
+        this.tokenEndpoint = httpEndpoint;
+    }
+
+    public doGenerate(): CSharpFile {
+        const class_ = csharp.class_({
+            ...this.classReference,
+            partial: true,
+            access: "public"
+        });
+
+        class_.addField(OauthTokenProviderGenerator.BUFFER_IN_MINUTES_FIELD);
+        class_.addField(OauthTokenProviderGenerator.ACCESS_TOKEN_FIELD);
+
+        if (this.getExpiresIn() != null) {
+            class_.addField(OauthTokenProviderGenerator.EXPIRES_AT_FIELD);
+        }
+
+        class_.addField(OauthTokenProviderGenerator.CLIENT_ID_FIELD);
+        class_.addField(OauthTokenProviderGenerator.CLIENT_SECRET_FIELD);
+
+        class_.addMethod(this.getAccessTokenMethod());
+
+        const subpackage = this.context.getSubpackageForServiceId(this.tokenEndpointReference.serviceId);
+        if (subpackage == null) {
+            throw new Error(`Failed to find subpackage for serviceId ${subpackage}`);
+        }
```

Comment:
Let's add an `OrThrow` variant on the `context` so that we never accidentally reimplement this elsewhere?

Reply:
def